torch_np/_detail/_reductions.py

```diff
@@ -0,0 +1,164 @@
+""" Implementation of reduction operations, to be wrapped into arrays, dtypes etc
+in the 'public' layer.
+
+Anything here only deals with torch objects, e.g. "dtype" is a torch.dtype instance etc
+"""
+
+import torch
+from . import _util
+from . import _scalar_types
+
+NoValue = None
+
+
+def _atleast_float(dtype, other_dtype):
+    """Return a dtype that is real or complex floating-point.
+
+    For inputs that are boolean or integer dtypes, this returns the default
+    float dtype; inputs that are complex get converted to the default complex
+    dtype; real floating-point dtypes (`float*`) get passed through unchanged
+    """
+    if dtype is None:
+        dtype = other_dtype
+    if not dtype.is_floating_point:
+        if dtype.is_complex:
+            pass  # pass through complex as is
+        else:
+            sctype = _scalar_types.default_float_type
+            dtype = sctype.torch_dtype
+    return dtype
+
+
+def count_nonzero(a, axis=None):
+    # XXX: this all should probably be generalized to a sum(a != 0, dtype=bool)
+    try:
+        return a.count_nonzero(axis)
+    except RuntimeError:
+        raise ValueError
+    return tensor
+
+
+def argmax(tensor, axis=None, out=None, *, keepdims=NoValue):
+    axis = _util.allow_only_single_axis(axis)
+    tensor = torch.argmax(tensor, axis)
+    return tensor
+
+def argmin(tensor, axis=None, out=None, *, keepdims=NoValue):
+    axis = _util.allow_only_single_axis(axis)
+    tensor = torch.argmin(tensor, axis)
+    return tensor
+
+
+def any(tensor, axis=None, *, where=NoValue):
+    if where is not None:
+        raise NotImplementedError
+
+    axis = _util.allow_only_single_axis(axis)
+
+    if axis is None:
+        result = tensor.any()
+    else:
+        result = tensor.any(axis)
+    return result
+
+
+def all(tensor, axis=None, *, where=NoValue):
+    if where is not None:
+        raise NotImplementedError
+
+    axis = _util.allow_only_single_axis(axis)
+
+    if axis is None:
+        result = tensor.all()
+    else:
+        result = tensor.all(axis)
+    return result
+
+
+def max(tensor, axis=None, initial=NoValue, where=NoValue):
+    if initial is not None or where is not None:
+        raise NotImplementedError
+
+    result = tensor.amax(axis)
+    return result
+
+
+def min(tensor, axis=None, initial=NoValue, where=NoValue):
+    if initial is not None or where is not None:
```

Comment:
Q. Should be testing here `is not None`  or `is not NoValue`?

Also, we should probably have one utils function for each of these common args.

Reply:
Thanks, changed. For now I kept the NoValue alias from numpy, not sure if it's really worth it. But changed for now anyway. 
Down the line these would certainly grow their dedicated utilities. Am going to push implementing `where` (via `torch.scatter` or some such) for after the initial coverage is more complete.

Reply:
We have `torch.where` and it's a particularly efficient function in PyTorch 2.0 in CUDA, so well worth checking that one.